core/state-machine/src/basic.rs

```diff
@@ -18,67 +18,67 @@
 
 use std::collections::HashMap;
 use std::iter::FromIterator;
+use crate::backend::{Backend, InMemory};
 use hash_db::Hasher;
 use trie::trie_root;
 use primitives::offchain;
-use primitives::storage::well_known_keys::{CHANGES_TRIE_CONFIG, CODE, HEAP_PAGES};
+use primitives::storage::well_known_keys::{HEAP_PAGES, is_child_storage_key};
 use parity_codec::Encode;
-use super::{ChildStorageKey, Externalities, OverlayedChanges};
+use super::{ChildStorageKey, Externalities};
 use log::warn;
 
 /// Simple HashMap-based Externalities impl.
+#[derive(Debug)]
 pub struct BasicExternalities {
-	inner: HashMap<Vec<u8>, Vec<u8>>,
-	changes: OverlayedChanges,
-	code: Option<Vec<u8>>,
+	top: HashMap<Vec<u8>, Vec<u8>>,
+	children: HashMap<Vec<u8>, HashMap<Vec<u8>, Vec<u8>>>,
 }
 
 impl BasicExternalities {
 	/// Create a new instance of `BasicExternalities`
-	pub fn new(inner: HashMap<Vec<u8>, Vec<u8>>) -> Self {
-		Self::new_with_code(&[], inner)
+	pub fn new(top: HashMap<Vec<u8>, Vec<u8>>) -> Self {
+		Self::new_with_children(top, Default::default())
 	}
 
-	/// Create a new instance of `BasicExternalities`
-	pub fn new_with_code(code: &[u8], mut inner: HashMap<Vec<u8>, Vec<u8>>) -> Self {
-		let mut overlay = OverlayedChanges::default();
-		super::set_changes_trie_config(
-			&mut overlay,
-			inner.get(&CHANGES_TRIE_CONFIG.to_vec()).cloned(),
-			false,
-		).expect("changes trie configuration is correct in test env; qed");
-
-		inner.insert(HEAP_PAGES.to_vec(), 8u64.encode());
+	/// Create a new instance of `BasicExternalities` with children
+	pub fn new_with_children(
+		mut top: HashMap<Vec<u8>, Vec<u8>>,
+		mut children: HashMap<Vec<u8>, HashMap<Vec<u8>, Vec<u8>>>,
+	) -> Self {
+		top.retain(|key, _| !is_child_storage_key(key));
+		children.retain(|key, _| is_child_storage_key(key));
```

Comment:
those removals weren't done in the past. Hmm I'm not sure they are needed as those checks will be done when writing to the actual externality no ?

Reply:
I'm not happy with this and the one in "real externalities". This could get someone in trouble when the genesis storage fucks something up, but it will not be found, because we accepted everything.

Reply:
"real externalities" is TestExternalities ?

maybe we can warn here as we do in `Ext` ?

Reply:
You called it `actual externality`, but I thought you speak about `Ext`.
I'm not sure a warning is enough. The question is, is this used in production code or just in tests and genesisconfig? If it is the later, I would maybe go with a panic to make the user aware that something is not correct.

Reply:
oh yes `actual externality` was `Ext`.

no it is used in production, ppl have started writing with_storage in add_genesis_config::build, and this with_storage makes use of basicExternality.

I changed TestExternality to panic when given wrong inputs and BasicExternality not do anything on wrong input (as when storages of genesis config will be put into Ext it will actually warn there I suppose).

I think we might want TestExternality panics, BasicExternality warns.

Reply:
I think we also want `BasicExternality` to panic. If there is a serious problem in your genesisconfig, you probably want to be aware of it. A warning that is maybe not seen, does not help you.